mlir/lib/Driver/Pipelines.cpp

```diff
@@ -0,0 +1,204 @@
+    pm.addPass(mlir::createReconcileUnrealizedCastsPass());
+    pm.addPass(catalyst::createGEPInboundsPass());
+    pm.addPass(catalyst::createRegisterInactiveCallbackPass());
+}
+
+void registerAllCatalystPipelines()
+{
+    registerEnforceRuntimeInvariantsPipeline();
+    registerHloLoweringPipeline();
+    registerQuantumCompilationPipeline();
+    registerBufferizationPipeline();
+    registerLLVMDialectLoweringPipeline();
+    registerDefaultCatalystPipeline();
+}
+
+void createDefaultCatalystPipeline(OpPassManager &pm)
+{
+    createEnforceRuntimeInvariantsPipeline(pm);
+    createHloLoweringPipeline(pm);
+    createQuantumCompilationPipeline(pm);
+    createBufferizationPipeline(pm);
+    createLLVMDialectLoweringPipeline(pm);
+}
+
+void registerEnforceRuntimeInvariantsPipeline()
+{
+    PassPipelineRegistration<>("enforce-runtime-invariants-pipeline",
+                               "Register enforce runtime invariants pipeline as a pass.",
+                               createEnforceRuntimeInvariantsPipeline);
+}
+void registerHloLoweringPipeline()
+{
+    PassPipelineRegistration<>("hlo_lowering-pipeline", "Register HLO lowering pipeline as a pass.",
+                               createHloLoweringPipeline);
+}
+void registerQuantumCompilationPipeline()
+{
+    PassPipelineRegistration<>("quantum-compilation-pipeline",
+                               "Register quantum compilation pipeline as a pass.",
+                               createQuantumCompilationPipeline);
+}
+void registerBufferizationPipeline()
+{
+    PassPipelineRegistration<>("bufferization-pipeline",
+                               "Register bufferization pipeline as a pass.",
+                               createBufferizationPipeline);
+}
+void registerLLVMDialectLoweringPipeline()
+{
+    PassPipelineRegistration<>("llvm-dialect-lowring-pipeline",
+                               "Register LLVM dialect lowring pipeline as a pass.",
+                               createLLVMDialectLoweringPipeline);
+}
+void registerDefaultCatalystPipeline()
+{
+    PassPipelineRegistration<>("default-catalyst-pipeline",
+                               "Register full default catalyst pipeline as a pass.",
+                               createDefaultCatalystPipeline);
+}
+
+std::vector<Pipeline> getDefaultPipeline()
+{
+    using PipelineFunc = void (*)(mlir::OpPassManager &);
+    std::vector<PipelineFunc> pipelineFuncs = {
+        &createEnforceRuntimeInvariantsPipeline, &createHloLoweringPipeline,
+        &createQuantumCompilationPipeline, &createBufferizationPipeline,
+        &createLLVMDialectLoweringPipeline};
+
+    Pipeline::PassList defaultPipelineNames = {
+        "enforce-runtime-invariants-pipeline", "hlo-lowering-pipeline",
+        "quantum-compilation-pipeline", "bufferization-pipeline", "llvm-dialect-lowering-pipeline"};
+
+    std::vector<Pipeline> defaultPipelines;
+    for (size_t i = 0; i < defaultPipelineNames.size(); ++i) {
+        Pipeline pipeline;
```

Comment:
Instead of making a bunch of `Pipeline` objects on the fly, `std::moving` them, and then destroying them when we exit the for loop's scope, consider the following alternative: since we know the length of the `defaultPipelines` vector, we can just initialize the vector with a fixed size and fill them with empty `Pipeline` objects. In the loop we simply set the fields.

(This would require the `Pipeline` class to have an empty constructor)

If it's too much effort to make it work, I'm also good with the current version : )

Reply:
The advantage is that now we can use `std::array` instead of `std::vector`, which should always be preferred when the "list size" is known.

Reply:
I didn't change use std::array since I have to [cast the result to a std::vector anyways ](https://github.com/PennyLaneAI/catalyst/blob/90edf551a49bb2132bbf051fcf04d885a474eee7/mlir/lib/Driver/CompilerDriver.cpp#L575).